src/env.h

```diff
@@ -824,7 +839,23 @@ class CleanupHookCallback {
   uint64_t insertion_order_counter_;
 };
 
-class Environment : public MemoryRetainer {
+// BindingDataBase serves as the base class for possible targets that can
+// be referred to from a FunctionTemplate's data field.
+class BindingDataBase : public BaseObject {
```

Comment:
We could do that, but I don’t really see a concrete advantage? It’s an extra object, yes, but it also gives us all the tools that come with `BaseObject`. It certainly keeps the code a lot simpler under the current requirements.

Reply:
>  It certainly keeps the code a lot simpler under the current requirements.

I am confused, how does this make the code simpler? We need to add an indirection with v8::External nonetheless, but with the usage of `BaseObject` we need to additionally create persistent `v8::Object`s for it. Without this, we could just unwrap and wrap the bindings directly.

> it also gives us all the tools that come with BaseObject

Which tools would be useful for the bindings? Adding references to it outside of the templates or or adding JS-land properties to it  would IMO just make the object graph more difficult to reason about. This also steers the templates further away from being context-independent, conceptually.

Reply:
> > It certainly keeps the code a lot simpler under the current requirements.
> 
> I am confused, how does this make the code simpler? We need to add an indirection with v8::External nonetheless, but with the usage of `BaseObject` we need to additionally create persistent `v8::Object`s for it.

There’s an extra object involved here, yes, but that comes at basically no cost?

> Without this, we could just unwrap and wrap the bindings through `v8::External` directly.

That is exactly what this PR does.

> > it also gives us all the tools that come with BaseObject
> 
> Which tools would be useful for the bindings? Adding references to it outside of the templates or or adding JS-land properties to it would IMO just make the object graph more difficult to reason about. This also steers the templates further away from being context-independent, conceptually.

We get the object lifetime management facilities, the snapshot serialization hooks, and the smart pointers provided by `BaseObject`, for starters, all of which are being used.

Reply:
I think the cost comes from the complexity of `BaseObject`s themselves, which doesn't seem necessary for something as simple as this.

> We get the object lifetime management facilities, the snapshot serialization hooks, and the smart pointers provided by BaseObject, for starters, all of which are being used.

Are we using the lifetime management or the smart pointers if there are no pure JS objects referencing these bindings, or pure JS objects being referenced by them? (we'd already need an additional weak callback for the persistent external handle no matter what). Also doesn't the serialization hook comes from other lower-level base class that the bindings can be based on directly? Are they serializing anything special other than the artifact of that `v8::Object`?

Reply:
@joyeecheung One thing we do currently use additionally is the fact that these objects get cleaned up by `FreeEnvironment()`, yes.

I’m not generally opposed to not using `BaseObject` here, but right now, this works, is relatively straightforward, and at least I personally don’t see how this could be simplified at the moment.

Reply:
> One thing we do currently use additionally is the fact that these objects get cleaned up by FreeEnvironment(), yes.

But we are already adding weak callbacks for the externals to call `BaseObject::MakeWeek`anyway? (For something we don't essentially need, I think, because there are no other references to them). If we don't use `BaseObject`s here we just need to delete the pointer to the binding in the weak callback, I'd think that's simpler compared to delegating the cleanup to an Environment (which is independent from GC).

Reply:
@joyeecheung In that case you’d manually have to add cleanup hooks. I think that comes with about the same complexity as the current code.

Reply:
@joyeecheung After thinking about this a bit more, I think the problem is not that we’re using `BaseObject`s for binding data, it’s really how we’re passing it.

The situation in which the current approach will eventually become problematic is when we actually start working on multi-instance (or, more specifically, multi-context) support for Node.js bindings. In that case, if we follow the current approach (and this is independent of this PR), we’ll create templates that are semantically context-dependent, and while I agree that that’s not great, it also doesn’t quite seem *that* bad to me.

One cleaner approach would be to stop using the `Data()` field of `FunctionCallbackInfo` to pass information about the binding/the `Environment` instance, and instead pass that information explicitly, for example as an argument to the function. I don’t see any way to implement that that wouldn’t mean having to perform a large-scale refactor of our code, but maybe we could do something like that on a per-module basis when we start working on multi-instance/multi-context support.

(I think all of this is out of scope for this PR, by the way.)

Reply:
To be exact, what do you think that needs to be wrapped into that `Data()`? The binding object?

If we are talking about grabbing bindings for the invocation context, I think we could just embed a chunk of context-independent metadata into `Data()`, so that in the callback

1) To get the Environment, we'd just use the Environment pointed embedded in the invocation context
2) To get the binding data, we get a numeric index from `Data()` to a list whose pointer is embedded into the Context. Then in the callback we just unwrap the binding index from `Data()`, and unwrap the list from `GetIsolate()->CurrentContext()->GetAlignedPointerFromEmbedderData(ContextEmbedderIndex::kBindings)`, then index into that list to retrieve the binding.

Reply:
@joyeecheung That sounds like a reasonable approach, yes.

> To be exact, what do you think that needs to be wrapped into that `Data()`? The binding object?

Nothing, we’d stop using it and instead pass the data explicitly as an argument (or as `this`) to the binding function. (The functions could be bound using `.bind()` to make that less awkward when they are returned from `internalBinding()`.) But that’s just an idea here, and I’m not saying that it’s better than what you’re suggesting.

Reply:
Hm, I guess if we simply stop using `Data()` it does require refactoring quite a lot of refactoring, but the riskier part is that we'd be changing the signatures/receivers of a lot of binding functions, some of which could be depended by...npm modules (sign)

Reply:
@joyeecheung Yeah, like I said, large-scale refactor ;)

At least for the time being, your suggestion does sound better. But, again, definitely out of scope of this PR imo.

Reply:
Why is it out of scope considering this PR adds `ExternalReferencePreAllocations` which is a result of this issue?

Reply:
@joyeecheung Because that issue exists either way. As long as we use externals, we have to pre-allocate whatever the `External`s that we serialize point to.

Reply:
But the issue for bindings should be solved differently if the inheritance doesn't exist? For example, `no_binding_data_deserializer` and `NoBindingData::Serialize` would be simplified because it then does not have states (the object index) that need to be serialized or deserialized. We just need to make sure the memory is pre-allocated so that V8 can hook the references back correctly.

Reply:
Which inheritance? The one from `BaseObject`?

I don’t think that makes a noticable difference here. The serialization code is pretty minimal anyway, and we still need to store which type the object has.